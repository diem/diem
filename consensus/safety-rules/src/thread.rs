// Copyright (c) The Libra Core Contributors
// SPDX-License-Identifier: Apache-2.0

//! This provides a execution separation between SafetyRules and Consensus without requiring the
//! use of processes. Rust does not support fork and so the mechanics to actually construct a
//! SafetyRules that would run together and be started by Consensus requires a separate binary and
//! making a call to start that via a command. This is a lightweight means of accomplishing a goal
//! in testing correctness of the communication layer between Consensus and SafetyRules.

use crate::{
    persistent_safety_storage::PersistentSafetyStorage,
    remote_service::{self, RemoteService},
};
use libra_config::utils;
use std::{
    net::{IpAddr, Ipv4Addr, SocketAddr},
    thread::{self, JoinHandle},
};

/// ThreadClient is the actual owner of the thread but in the context of Consenus and SafetyRules
/// is on the client side of the operations as it makes queries / requests to SafetyRules.
pub struct ThreadService {
    _child: JoinHandle<()>,
    server_addr: SocketAddr,
    network_timeout: u64,
}

impl ThreadService {
    pub fn new(
        storage: PersistentSafetyStorage,
        verify_vote_proposal_signature: bool,
        timeout: u64,
    ) -> Self {
        let listen_port = utils::get_available_port();
        let listen_addr = SocketAddr::new(IpAddr::V4(Ipv4Addr::LOCALHOST), listen_port);
        let server_addr = listen_addr;

        let child = thread::spawn(move || {
            remote_service::execute(
                storage,
                listen_addr,
                verify_vote_proposal_signature,
                timeout,
            )
        });

        Self {
            _child: child,
            server_addr,
            network_timeout: timeout,
        }
    }
}

impl RemoteService for ThreadService {
    fn server_address(&self) -> SocketAddr {
        self.server_addr
    }
    fn network_timeout_ms(&self) -> u64 {
        self.network_timeout
    }
}
