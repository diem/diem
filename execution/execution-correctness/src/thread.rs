// Copyright (c) The Libra Core Contributors
// SPDX-License-Identifier: Apache-2.0

//! This provides a runtime separation between ExecutionCorrectness and the rest without requiring the
//! use of processes. Rust does not support fork and so the mechanics to actually construct a
//! ExecutionCorrectness that would run together and be started by SafetyRules requires a separate binary and
//! making a call to start that via a command. This is a lightweight means of accomplishing a goal
//! in testing correctness of the communication layer between ExecutionCorrectness and SafetyRules.

use crate::remote_service::{self, RemoteService};
use libra_config::utils;
use libra_crypto::ed25519::Ed25519PrivateKey;
use std::{
    net::{IpAddr, Ipv4Addr, SocketAddr},
    thread::{self, JoinHandle},
};

/// ThreadClient is the actual owner of the thread but in the context of SafetyRules and ExecutionCorrectness
/// is on the client side of the operations as it makes queries / requests to ExecutionCorrectness.
pub struct ThreadService {
    _child: JoinHandle<()>,
    server_addr: SocketAddr,
    network_timeout: u64,
}

impl ThreadService {
    pub fn new(
        storage_addr: SocketAddr,
        prikey: Option<Ed25519PrivateKey>,
        network_timeout: u64,
    ) -> Self {
        let listen_port = utils::get_available_port();
        let listen_addr = SocketAddr::new(IpAddr::V4(Ipv4Addr::LOCALHOST), listen_port);
        let server_addr = listen_addr;

        let child = thread::spawn(move || {
            remote_service::execute(storage_addr, listen_addr, prikey, network_timeout)
        });

        Self {
            _child: child,
            server_addr,
            network_timeout,
        }
    }
}

impl RemoteService for ThreadService {
    fn server_address(&self) -> SocketAddr {
        self.server_addr
    }
    fn network_timeout(&self) -> u64 {
        self.network_timeout
    }
}
