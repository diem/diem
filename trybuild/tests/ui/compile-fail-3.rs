use std::ptr;
use std::thread;

fn main() {
    let x = ptr::null_mut();

    thread::spawn(|| {
        println!("{:?}", x)
    });
}
