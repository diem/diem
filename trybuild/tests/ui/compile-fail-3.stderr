error[E0277]: `*mut _` cannot be shared between threads safely
   --> $DIR/compile-fail-3.rs:7:5
    |
7   |     thread::spawn(|| {
    |     ^^^^^^^^^^^^^ `*mut _` cannot be shared between threads safely
    |
    = help: the trait `std::marker::Sync` is not implemented for `*mut _`
    = note: required because of the requirements on the impl of `std::marker::Send` for `&*mut _`
    = note: required because it appears within the type `[closure@$DIR/tests/ui/compile-fail-3.rs:7:19: 9:6 x:&*mut _]`
